{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hamming\n",
    "\n",
    "Calculate the Hamming difference between two DNA strands.\n",
    "\n",
    "A mutation is simply a mistake that occurs during the creation or\n",
    "copying of a nucleic acid, in particular DNA. Because nucleic acids are\n",
    "vital to cellular functions, mutations tend to cause a ripple effect\n",
    "throughout the cell. Although mutations are technically mistakes, a very\n",
    "rare mutation may equip the cell with a beneficial attribute. In fact,\n",
    "the macro effects of evolution are attributable by the accumulated\n",
    "result of beneficial microscopic mutations over many generations.\n",
    "\n",
    "The simplest and most common type of nucleic acid mutation is a point\n",
    "mutation, which replaces one base with another at a single nucleotide.\n",
    "\n",
    "By counting the number of differences between two homologous DNA strands\n",
    "taken from different genomes with a common ancestor, we get a measure of\n",
    "the minimum number of point mutations that could have occurred on the\n",
    "evolutionary path between the two strands.\n",
    "\n",
    "This is called the 'Hamming distance'.\n",
    "\n",
    "It is found by comparing two DNA strands and counting how many of the\n",
    "nucleotides are different from their equivalent in the other string.\n",
    "\n",
    "    GAGCCTACTAACGGGAT\n",
    "    CATCGTAATGACGGCCT\n",
    "    ^ ^ ^  ^ ^    ^^\n",
    "\n",
    "The Hamming distance between these two DNA strands is 7.\n",
    "\n",
    "# Implementation notes\n",
    "\n",
    "The Hamming distance is only defined for sequences of equal length, so\n",
    "an attempt to calculate it between sequences of different lengths should\n",
    "not work. The general handling of this situation (e.g., raising an\n",
    "exception vs returning a special value) may differ between languages.\n",
    "## Source\n",
    "\n",
    "The Calculating Point Mutations problem at Rosalind [http://rosalind.info/problems/hamm/](http://rosalind.info/problems/hamm/)\n",
    "\n",
    "\n",
    "## Version compatibility\n",
    "Julia 1.0 and 0.7 are the only supported Julia versions on Exercism.\n",
    "For the most part, the test suites and solutions should be compatible to 0.6, but you will have to change `using Test` back to `using Base.Test` in the `runtests.jl` file.\n",
    "Note that 0.7 and 1.0 are almost identical, except for deprecation warnings, which have all been removed in 1.0.\n",
    "\n",
    "## Submitting Incomplete Solutions\n",
    "It's possible to submit an incomplete solution so you can see how others have completed the exercise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# submit\n",
    "function distance(s1::AbstractString, s2::AbstractString)\n",
    "\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Test\n",
    "\n",
    "# include(\"hamming.jl\")\n",
    "\n",
    "@testset \"identical strands\" begin\n",
    "    @test distance(\"A\", \"A\") == 0\n",
    "    @test distance(\"GGACTGA\", \"GGACTGA\") == 0\n",
    "end\n",
    "\n",
    "@testset \"complete distance\" begin\n",
    "    @test distance(\"A\", \"G\") == 1\n",
    "    @test distance(\"AG\", \"CT\") == 2\n",
    "end\n",
    "\n",
    "@testset \"small distance\" begin\n",
    "    @test distance(\"AT\", \"CT\") == 1\n",
    "    @test distance(\"GGACG\", \"GGTCG\") == 1\n",
    "    @test distance(\"ACCAGGG\", \"ACTATGG\") == 2\n",
    "end\n",
    "\n",
    "@testset \"non-unique characters\" begin\n",
    "    @test distance(\"AGA\", \"AGG\") == 1\n",
    "    @test distance(\"AGG\", \"AGA\") == 1\n",
    "end\n",
    "\n",
    "@testset \"same nucleotides in different positions\" begin\n",
    "    @test distance(\"TAG\", \"GAT\") == 2\n",
    "end\n",
    "\n",
    "@testset \"large distance\" begin\n",
    "    @test distance(\"GATACA\", \"GCATAA\") == 4\n",
    "    @test distance(\"GGACGGATTCTG\", \"AGGACGGATTCT\") == 9\n",
    "end\n",
    "\n",
    "@testset \"empty strands\" begin\n",
    "    @test distance(\"\", \"\") == 0\n",
    "end\n",
    "\n",
    "@testset \"different strand lengths\" begin\n",
    "    @test_throws ArgumentError distance(\"AATG\", \"AAA\")\n",
    "    @test_throws ArgumentError distance(\"ATA\", \"AGTG\")\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To submit your exercise, you need to save your solution in a file called hamming.jl before using the CLI.\n",
    "# You can either create it manually or use the following functions, which will automatically\n",
    "# save every notebook cell starting with `# submit` in that file.\n",
    "\n",
    "# Pkg.add(\"Exercism\")\n",
    "# using Exercism\n",
    "# Exercism.create_submission(\"hamming\")\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
